source/common/router/rds_impl.cc

```diff
@@ -119,5 +125,45 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
   init_manager.registerTarget(*this);
 }
 
+RouteConfigProviderManagerImpl::RouteConfigProviderManagerImpl(
+    Runtime::Loader& runtime, Event::Dispatcher& dispatcher, Runtime::RandomGenerator& random,
+    const LocalInfo::LocalInfo& local_info, ThreadLocal::SlotAllocator& tls)
+    : runtime_(runtime), dispatcher_(dispatcher), random_(random), local_info_(local_info),
+      tls_(tls) {}
+
+std::vector<Router::RouteConfigProviderSharedPtr>
+RouteConfigProviderManagerImpl::routeConfigProviders() {
+  std::vector<Router::RouteConfigProviderSharedPtr> ret;
+  ret.reserve(route_config_providers_.size());
+  for (const auto& element : route_config_providers_) {
+    ret.push_back(element.second.lock());
+  }
+  return ret;
+};
+
+Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::getRouteConfigProvider(
+    const Json::Object& config, Upstream::ClusterManager& cm, Stats::Scope& scope,
+    const std::string& stat_prefix, Init::Manager& init_manager) {
+
+  // RdsRouteConfigProviders are unique based on their <route_config_name>_<cluster>.
+  std::string map_identifier = config.getString("route_config_name") + "_";
+  map_identifier += config.getString("cluster");
+
+  auto it = route_config_providers_.find(map_identifier);
+  if (it == route_config_providers_.end()) {
+    std::shared_ptr<RdsRouteConfigProviderImpl> new_provider{new RdsRouteConfigProviderImpl(
+        config, runtime_, cm, dispatcher_, random_, local_info_, scope, stat_prefix, tls_, *this)};
+
+    new_provider->registerInitTarget(init_manager);
+
+    route_config_providers_.insert(
+        {map_identifier, std::weak_ptr<RdsRouteConfigProviderImpl>(new_provider)});
+
+    return std::move(new_provider);
+  }
+
+  return it->second.lock();
```

Comment:
Can this `lock()` fail? Based on other comment threads and the comment above about only cleaning up in `RouteConfigProvider` destructors, I'm thinking no. If that's the case, please add an ASSERT/comment to document. 

If it can fail here, shouldn't we have similar handling to the "not found" case above and create a new provider?